### ISSUE TITLE
Times out?

### ISSUE THREAD

**Issue by JohnInWI** [gilmaimon/ArduinoWebsockets#20]
Nice library & easy to use.
I have a websocket "bridge" (node-red) set up on a remote server, so I'm using the example client sketch on a ESP8266.
The ESP connects to the remote socket and works well, but after a period of time and inactivity, disconnects. If I re-boot the ESP it connects again. The remote server is not the cause, as other sockets remain connected indefinitely.

I'm new to the websocket world, so perhaps I'm overlooking the obvious? or, is there a built-in timeout in the library i might be able to change?

**Comment by gilmaimon** [comment]
Hi, thank you for opening an issue.

I'm not sure exactly if the server sends the disconnect or the server. If you could turn on logs (on highest verbosity) it could be very helpfull.

Anyway, the best way is to send some kind of "keepalive" message to the server, which is exactly why the websockets protocol has Ping packets.

Try adding code which sends a ping to the server every few minutes or so, it should fix your issue completely. (Just make sure you don't spam the server with pings without any delay).

I would still appreciate logs, and your issue makes me consider adding keep-alive pings as library feature. Let me know if it is something you are interested in :)

I'm on mobile so I apologise for not adding code samples. Let me know if you need any extra info or help!

Thanks again for using the library!

Have fun,
Gil.

**Comment by adelin-mcbsoft** [comment]
Hey,

I would like to have add my opinion too here:

Just in case you consider adding _keep-alive pings_ in the future, please make sure you add an optional flag for that (so you could set if you want them or not), as there are server/client implementations (see Web Browsers for example) which do not support ping / pong(s).

(Web Browsers in particular, act pretty ugly on them, they disconnect when they receive a ping/pong).

But anyway - may be a good idea to have them.

Regarding @JohnInWI's issue, I experience some sort of disconnects even after long-heavy-usage, but is still under investigation, the library might not be at fault, may be a HTTP SSL problem or so.

'til we find out - thank you once again Gil for the library! Best in town :D !

Have a great day,
A.

**Comment by gilmaimon** [comment]
So, I consider this a potential feature rather than a bug. On my TODO list.

Gil.

**Comment by JohnInWI** [comment]
Agreed- a **feature** rather than a bug.  I can add a periodic ping to keep connection alive. The way it is, killing a non-active socket might be useful in some situations... although I cannot confirm it is the library doing the disconnect at this point.
Not sure how to get a detailed log from Node-Red's websockets... working on that... (no time!)
Again, thanks for the great library.

**Comment by gilmaimon** [comment]
If you are using an ESP32, you can see the logging settings in the board menu. It's better to get the logs form the board itself rather than from the server.

For me, adding periodic pings fixed simmilar issue before (usually it is server configuration to avoid zombie connections). 

Gil.

**Comment by gilmaimon** [comment]
So, I thought about this and my opinion is that a periodic ping is not something that should be implemented inside the client itself. It bounds the usage of the client to `.poll` style of usage (instead of a readBlocking, if some users want it) and overall its a responsibility I don't want the client to have.

There are other solutions for doing work on fixed intervals, one of them is in my [ArduinoComponents](https://github.com/gilmaimon/ArduinoComponents) library (but if you are reading this you should probably find a better maintained library).

I'm closing the issue, but it is still open for discussion, I'm available via email mail.gilmaimon@gmail.com
Gil.
